### ISSUE TITLE
callback preview image noise

### ISSUE THREAD

**Issue by enescakircali** [huggingface/diffusers#3579]
```py
import torch
import torchvision
from PIL import Image
from diffusers import EulerDiscreteScheduler, StableDiffusionPipeline

device = "cuda" if torch.cuda.is_available() else "cpu"

pipe = StableDiffusionPipeline.from_pretrained(
    "prompthero/openjourney-v4", torch_dtype=torch.float16, safety_checker=None)
pipe = pipe.to(device)
pipe.enable_attention_slicing()
pipe.scheduler = EulerDiscreteScheduler.from_config(
    pipe.scheduler.config, use_karras_sigmas=True
)

prompt = "Turkish"

negative_prompt = "low quality"

# num_images_per_prompt=4,

def progress(step, timestep, latents):
    print(step, timestep, latents[0][0][0][0])

    with torch.no_grad():

        latents = 1 / 0.18215 * latents
        image = pipe.vae.decode(latents).sample

        image = (image / 2 + 0.5).clamp(0, 1)

        # we always cast to float32 as this does not cause significant overhead and is compatible with bfloa16
        image = image.cpu().permute(0, 2, 3, 1).float().numpy()

        # convert to PIL Images
        image = pipe.numpy_to_pil(image)

        # do something with the Images
        for i, img in enumerate(image):
            img.save(f"step_{step}_img{i}.png")


result = pipe(prompt=prompt,
              num_inference_steps=20,
              height=512, width=512,
              guidance_scale=7,
              negative_prompt=negative_prompt,
              callback=progress,
              callback_steps=5
              )

image = result.images[0]

image.save(f"outputs/cikar goster.png")
print(result.nsfw_content_detected)

 ```
 
I have such a code where I'm saving a progress image step by step in the callback. When saving, it looks like this 

https://github.com/huggingface/diffusers/assets/31443145/0e153756-e712-4fc9-a6ac-6324708d3696

but I want it to look like this: 

https://github.com/huggingface/diffusers/assets/31443145/0931dd67-4e8d-4095-85c3-512ad80e4060

 How can I achieve this?

**Comment by Abhinay1997** [comment]
@enescakircali, the default output type is 'pil'. So you don't need to do vae decoding or anything else in your callback. Try this code:

```python
import torch
import torchvision
from PIL import Image
from diffusers import EulerDiscreteScheduler, StableDiffusionPipeline

device = "cuda" if torch.cuda.is_available() else "cpu"

pipe = StableDiffusionPipeline.from_pretrained(
    "prompthero/openjourney-v4", torch_dtype=torch.float16, safety_checker=None)
pipe = pipe.to(device)
pipe.enable_attention_slicing()
pipe.scheduler = EulerDiscreteScheduler.from_config(
    pipe.scheduler.config, use_karras_sigmas=True
)

prompt = "Turkish woman holding an umbrella"

negative_prompt = "low quality"

# num_images_per_prompt=4,

def progress(step, timestep, latents):
    print(step, timestep, latents[0][0][0][0])

    with torch.no_grad():

        latents = 1 / 0.18215 * latents
        image = pipe.vae.decode(latents).sample

        image = (image / 2 + 0.5).clamp(0, 1)

        # we always cast to float32 as this does not cause significant overhead and is compatible with bfloa16
        image = image.cpu().permute(0, 2, 3, 1).float().numpy()

        # convert to PIL Images
        image = pipe.numpy_to_pil(image)

        # do something with the Images
        for i, img in enumerate(image):
            img.save(f"step_{step}_img{i}.png")


#You just need the output_type = 'latent' for this to work. 
result = pipe(prompt=prompt,
              num_inference_steps=20,
              height=512, width=512,
              guidance_scale=7,
              negative_prompt=negative_prompt,
              callback=progress,
              callback_steps=1,
              output_type='latent'
              )

```

My video:
I couldn't reduce the framerate below 24fps with my video editor, but increase the `num_inference_steps` to 30 maybe and keep the `callback_steps` = 1 for better video.

https://github.com/huggingface/diffusers/assets/24771261/f46908f7-fe57-40f5-a9d9-1516411f8ec9

**Comment by KiddoGeeGee** [comment]
same question.

**Comment by enescakircali** [comment]
```py

import torch, logging
## disable warnings
#logging.disable(logging.WARNING)
## Imaging  library
from PIL import Image
from torchvision import transforms as tfms
## Basic libraries
import numpy as np
from tqdm.auto import tqdm
import matplotlib.pyplot as plt
#matplotlib inline
from IPython.display import display
import shutil
import os
## For video display
from IPython.display import HTML
from base64 import b64encode

import discord
from io import BytesIO

## Import the CLIP artifacts 
from transformers import CLIPTextModel, CLIPTokenizer
from diffusers import AutoencoderKL, UNet2DConditionModel, LMSDiscreteScheduler
## Initiating tokenizer and encoder.
tokenizer = CLIPTokenizer.from_pretrained("openai/clip-vit-large-patch14", torch_dtype=torch.float16)
text_encoder = CLIPTextModel.from_pretrained("openai/clip-vit-large-patch14", torch_dtype=torch.float16).to("cuda")
## Initiating the VAE
vae = AutoencoderKL.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="vae", torch_dtype=torch.float16).to("cuda")
## Initializing a scheduler and Setting number of sampling steps
scheduler = LMSDiscreteScheduler(beta_start=0.00085, beta_end=0.012, beta_schedule="scaled_linear", num_train_timesteps=1000)
scheduler.set_timesteps(50)
## Initializing the U-Net model
unet = UNet2DConditionModel.from_pretrained("CompVis/stable-diffusion-v1-4", subfolder="unet", torch_dtype=torch.float16).to("cuda")
## Helper functions
def load_image(p):
    '''
    Function to load images from a defined path
    '''
    return Image.open(p).convert('RGB').resize((512,512))
def pil_to_latents(image):
    '''
    Function to convert image to latents
    '''
    init_image = tfms.ToTensor()(image).unsqueeze(0) * 2.0 - 1.0
    init_image = init_image.to(device="cuda", dtype=torch.float16) 
    init_latent_dist = vae.encode(init_image).latent_dist.sample() * 0.18215
    return init_latent_dist
def latents_to_pil(latents):
    '''
    Function to convert latents to images
    '''
    latents = (1 / 0.18215) * latents
    with torch.no_grad():
        image = vae.decode(latents).sample
    image = (image / 2 + 0.5).clamp(0, 1)
    image = image.detach().cpu().permute(0, 2, 3, 1).numpy()
    images = (image * 255).round().astype("uint8")
    pil_images = [Image.fromarray(image) for image in images]
    return pil_images
def text_enc(prompts, maxlen=None):
    '''
    A function to take a texual promt and convert it into embeddings
    '''
    if maxlen is None: maxlen = tokenizer.model_max_length
    inp = tokenizer(prompts, padding="max_length", max_length=maxlen, truncation=True, return_tensors="pt") 
    return text_encoder(inp.input_ids.to("cuda"))[0].half()
async def prompt_2_img(prompt, g=7.5, batch_size=1, seed=0, steps=70, height=512, width=512, message=None, context=None):
    """
    Diffusion process to convert prompt to image
    """
    warning_step = 0
    
    # Converting textual prompts to embedding
    text = text_enc([prompt] * batch_size)
    
    # Adding an unconditional prompt , helps in the generation process
    uncond =  text_enc([""] * batch_size, text.shape[1])
    emb = torch.cat([uncond, text])
    
    # Setting the seed
    if seed: torch.manual_seed(seed)
    
    # Initiating random noise
    latents = torch.randn((batch_size, unet.in_channels, height//8, width//8))
    
    # Setting number of steps in scheduler
    scheduler.set_timesteps(steps)
    
    # Adding noise to the latents 
    latents = latents.to("cuda").half() * scheduler.init_noise_sigma
    
    # Iterating through defined steps
    for i,ts in enumerate(tqdm(scheduler.timesteps)):
        # We need to scale the i/p latents to match the variance
        inp = scheduler.scale_model_input(torch.cat([latents] * 2), ts)
        
        # Predicting noise residual using U-Net
        with torch.no_grad(): u,t = unet(inp, ts, encoder_hidden_states=emb).sample.chunk(2)
            
        # Performing Guidance
        pred = u + g*(t-u)
        
        # Conditioning  the latents
        latents = scheduler.step(pred, ts, latents).prev_sample
        
        # Saving intermediate images
        if context.author and message.content:
            if warning_step == i:
               images =  latents_to_pil(latents)
               canvas = Image.new('RGBA', (width*2, height*2), (255, 255, 255, 0))
               if len(images) >= 1:
                    canvas.paste(images[0], (0, 0))
               if len(images) >= 2:
                    canvas.paste(images[1], (width, 0))
               if len(images) >= 3:
                    canvas.paste(images[2], (0, height))
               if len(images) == 4:
                    canvas.paste(images[3], (width, height))
               image_bytes = BytesIO()
               # Save the image data to the BytesIO object
               canvas.save(image_bytes, format='PNG')
               image_bytes.seek(0)
               file = discord.File(image_bytes, filename=f"steps_{i:04}.png")
               await message.edit(content=f"**{prompt}** - {context.author.mention} ({i%steps}%) ({'fast' if steps <= 70 else 'slow'})", attachments=[file])
               warning_step += 10
            
    # Returning the latent representation to output an image of 3x512x512
    return latents_to_pil(latents)
```

by the way i already changed those codes

Color noise, which is not very cool because I don't want the uploaded image stepper to have noise. (according to me)

**Comment by enescakircali** [comment]
It would be great if you had a more detailed document.

**Comment by enescakircali** [comment]
```py
# Initiating random noise
    latents = torch.randn((batch_size, unet.in_channels, height//8, width//8))
```
Is there any other way to do this?
as in the 2nd video

**Comment by CrazyBoyM** [comment]
same question.

**Comment by skirsten** [comment]
Hi, I am also trying to generate preview images using something similar to https://github.com/madebyollin/taesd/blob/main/examples/Previewing_During_Image_Generation.ipynb.

Unfortunately the formula from this notebook does not work well in the beginning of the discrete schedulers.
## Non discrete (works great)

### DDIMScheduler
![DDIMScheduler webp](https://github.com/huggingface/diffusers/assets/1972314/41f5e99e-f7d6-417e-b671-d5bc92805644)
### DDPMScheduler
![DDPMScheduler webp](https://github.com/huggingface/diffusers/assets/1972314/e6528242-1e5e-4de4-a191-70d9f0df96ef)
### DPMSolverMultistepScheduler
![DPMSolverMultistepScheduler webp](https://github.com/huggingface/diffusers/assets/1972314/cebd9cc4-06c7-4963-81d1-2aa97d14d034)
### PNDMScheduler
![PNDMScheduler webp](https://github.com/huggingface/diffusers/assets/1972314/de0bcb36-2d8c-4391-82a2-f28ffaf2684a)

## Discrete (not so great)

### EulerAncestralDiscreteScheduler
![EulerAncestralDiscreteScheduler webp](https://github.com/huggingface/diffusers/assets/1972314/2cca4c49-c82a-4998-be32-e9a491a9c9b5)
### EulerDiscreteScheduler
![EulerDiscreteScheduler webp](https://github.com/huggingface/diffusers/assets/1972314/9e7dac17-dc49-4155-bc82-ff42d20312ef)
### LMSDiscreteScheduler
![LMSDiscreteScheduler webp](https://github.com/huggingface/diffusers/assets/1972314/2f7557fa-9cb2-41ad-ad9f-ae4d5540ecbf)

This issue is the only thing I could find that is related to preview images and schedulers. Anybody knows if its possible to change the formula to work for discrete schedulers? Any help is appreciated :)

**Comment by madebyollin** [comment]
Hmm, I think there's an underlying limitation in the pipeline code right now. `pred_original_sample` is [computed](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_ddim.py#L467) in the scheduler `step` but currently [discarded](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L695C113-L695C113) by the pipeline, so pipeline callbacks are restricted to only previewing `prev_sample` (instead of the `pred_original_sample` which you'd typically want to preview).

To fix, we'd need to modify the pipeline code to keep `pred_original_sample` around and pass it to the callback:
```diff
# compute the previous noisy sample x_t -> x_t-1
- latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs, return_dict=False)[0]
+ # and the predicted original sample (in case callbacks want to preview it)
+ latents, pred_original_sample = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs)

# call the callback, if provided
if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
    progress_bar.update()
    if callback is not None and i % callback_steps == 0:
-        callback(i, t, latents)
+        callback(i, t, latents, pred_original_sample=pred_original_sample)
```

A change like this would allow users to preview `pred_original_sample` using pipeline callbacks instead of [ugly hacks](https://github.com/madebyollin/taesd/blob/main/examples/Previewing_During_Image_Generation.ipynb), although I suspect further edits would be needed to avoid breaking `torch.compile`.

**Comment by kopyl** [comment]
> Hmm, I think there's an underlying limitation in the pipeline code right now. `pred_original_sample` is [computed](https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_ddim.py#L467) in many scheduler `step`s but currently [discarded](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L695C113-L695C113) by the pipeline, so pipeline callbacks are restricted to only previewing `prev_sample` (instead of the `pred_original_sample` which you'd typically want to preview).
> 
> To fix, we'd need to modify the pipeline code to keep `pred_original_sample` around and pass it to the callback:
> 
> ```diff
> # compute the previous noisy sample x_t -> x_t-1
> - latents = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs, return_dict=False)[0]
> + # and the predicted original sample (in case callbacks want to preview it)
> + latents, pred_original_sample = self.scheduler.step(noise_pred, t, latents, **extra_step_kwargs)
> 
> # call the callback, if provided
> if i == len(timesteps) - 1 or ((i + 1) > num_warmup_steps and (i + 1) % self.scheduler.order == 0):
>     progress_bar.update()
>     if callback is not None and i % callback_steps == 0:
> -        callback(i, t, latents)
> +        callback(i, t, latents, pred_original_sample=pred_original_sample)
> ```
> 
> A change like this would allow users to preview `pred_original_sample` using pipeline callbacks instead of [ugly hacks](https://github.com/madebyollin/taesd/blob/main/examples/Previewing_During_Image_Generation.ipynb), although I think further edits would be needed to avoid breaking `torch.compile` (and to compute `pred_original_sample` within _every_ scheduler).

Thank you ❤️ 
That helped.
Although i realized that [Flux scheduler's step method](https://github.com/huggingface/diffusers/blob/dac623b59f52c58383a39207d5147aa34e0047cd/src/diffusers/schedulers/scheduling_flow_match_euler_discrete.py#L242) does not return `pred_original_sample`, so I had to add it myself like this:

```
gamma = 0
sigma_hat = sigma * (gamma + 1)
sigma_input = sigma_hat
pred_original_sample = model_output * (-sigma_input / (sigma_input**2 + 1) ** 0.5) + (sample / (sigma_input**2 + 1))
```
(you have to add it after the line with `self._step_index += 1`
And then i had to modify the [Flux's pipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux.py) like this:
Replace `latents = self.scheduler.step(noise_pred, t, latents, return_dict=False)` with `latents, pred_original_sample = self.scheduler.step(noise_pred, t, latents, return_dict=False)`
and replace `callback_outputs = callback_on_step_end(self, i, t, callback_kwargs)` with `callback_outputs = callback_on_step_end(self, i, t, callback_kwargs, pred_original_sample=pred_original_sample)`
The same also works for the [Img2Img pipeline](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux_img2img.py).
